File: extensions/broker-server-postgres-flyway-jooq/src/main/java/de/sovity/edc/ext/brokerserver/db/FlywayFactory.java
/*
 *  Copyright (c) 2023 sovity GmbH
 *
 *  This program and the accompanying materials are made available under the
 *  terms of the Apache License, Version 2.0 which is available at
 *  https://www.apache.org/licenses/LICENSE-2.0
 *
 *  SPDX-License-Identifier: Apache-2.0
 *
 *  Contributors:
 *       sovity GmbH - initial implementation
 *
 */

package de.sovity.edc.ext.brokerserver.db;

import lombok.RequiredArgsConstructor;
import org.eclipse.edc.spi.system.configuration.Config;
import org.flywaydb.core.Flyway;

import javax.sql.DataSource;

import static de.sovity.edc.ext.brokerserver.db.PostgresFlywayExtension.FLYWAY_CLEAN_ENABLE;

/**
 * Quickly launch {@link Flyway} from EDC Config
 */
@RequiredArgsConstructor
public class FlywayFactory {
    private final Config config;

    /**
     * Configure and launch {@link Flyway}.
     *
     * @param dataSource data source
     * @return {@link Flyway}
     */
    public Flyway setupFlyway(DataSource dataSource) {
        return Flyway.configure()
                .baselineOnMigrate(true)
                .dataSource(dataSource)
                .cleanDisabled(!config.getBoolean(FLYWAY_CLEAN_ENABLE, false))
                .table("flyway_schema_history")
                .locations("classpath:db/migration")
                .load();
    }
}

File: connector/ce/connector-ce/src/main/kotlin/de/sovity/edc/ce/modules/db/FlywayFactory.kt
/*
 * Copyright sovity GmbH and/or licensed to sovity GmbH under one or
 * more contributor license agreements. You may not use this file except
 * in compliance with the "Elastic License 2.0".
 *
 * SPDX-License-Identifier: Elastic-2.0
 */
package de.sovity.edc.ce.modules.db

import de.sovity.edc.ce.config.CeConfigProps
import de.sovity.edc.ce.modules.db.DbExtension.JdbcCredentials
import org.eclipse.edc.spi.system.configuration.Config
import org.flywaydb.core.Flyway

object FlywayFactory {
    fun newFlyway(config: Config, jdbcCredentials: JdbcCredentials): Flyway {
        val locations = mutableListOf<String>()
        locations.add(CeConfigProps.SOVITY_FLYWAY_MIGRATION_LOCATION.getStringOrThrow(config))
        locations.addAll(CeConfigProps.SOVITY_FLYWAY_ADDITIONAL_MIGRATION_LOCATIONS.getListOrEmpty(config))

        return Flyway.configure()
            .dataSource(jdbcCredentials.url, jdbcCredentials.user, jdbcCredentials.password)
            .cleanDisabled(!CeConfigProps.SOVITY_FLYWAY_CLEAN_ENABLE.getBooleanOrFalse(config))
            .baselineVersion("0")
            .baselineOnMigrate(true)
            .table("flyway_schema_history_sovity")
            .locations(*locations.toTypedArray<String?>())
            .load()
    }
}
